taskiq/middlewares/opentelemetry_middleware.py

```diff
@@ -0,0 +1,299 @@
+    we end up trying to publish a task with the same id as the task currently running.
+    """
+    if message is None:
+        return
+
+    ctx_dict = getattr(message, CTX_KEY, None)
+
+    if ctx_dict is None:
+        ctx_dict = {}
+        setattr(message, CTX_KEY, ctx_dict)
+
+    ctx_dict[(message.task_id, is_publish)] = (span, activation, token)
+
+
+def detach_context(message: TaskiqMessage, is_publish: bool = False) -> None:
+    """Remove context from `TaskiqMessage`."""
+    span_dict = getattr(message, CTX_KEY, None)
+    if span_dict is None:
+        return
+
+    # See note in `attach_context` for key info
+    span_dict.pop((message.task_id, is_publish), None)
+
+
+def retrieve_context(
+    message: TaskiqMessage,
+    is_publish: bool = False,
+) -> Optional[Tuple[Span, AbstractContextManager[Span], Optional[object]]]:
+    """Retrieve context from `TaskiqMessage`."""
+    span_dict = getattr(message, CTX_KEY, None)
+    if span_dict is None:
+        return None
+
+    # See note in `attach_context` for key info
+    return span_dict.get((message.task_id, is_publish), None)
+
+
+class OpenTelemetryMiddleware(TaskiqMiddleware):
+    """Middleware to instrument Taskiq with OpenTelemetry."""
+
+    def __init__(
+        self,
+        tracer_provider: Optional[TracerProvider] = None,
+        meter_provider: Optional[MeterProvider] = None,
+        tracer: Optional[Tracer] = None,
+        meter: Optional[Meter] = None,
+    ) -> None:
+        super().__init__()
+        self._tracer = (
+            trace.get_tracer(
+                __name__,
+                __version__,
+                tracer_provider,
+                schema_url="https://opentelemetry.io/schemas/1.11.0",
+            )
+            if tracer is None
+            else tracer
+        )
+        self._meter = (
+            get_meter(
+                __name__,
+                __version__,
+                meter_provider,
+                schema_url="https://opentelemetry.io/schemas/1.11.0",
+            )
+            if meter is None
+            else meter
+        )
+
+    def pre_send(self, message: TaskiqMessage) -> TaskiqMessage:
+        """
+        This function starts new span and propagates opentelemetry state to labels.
+
+        :param message: current message.
+        :return: message
+        """
+        logger.debug("pre_send task_id=%s", message.task_id)
+
+        operation_name = f"{_TASK_SEND}/{message.task_name}"
+        span = self._tracer.start_span(operation_name, kind=trace.SpanKind.PRODUCER)
+
+        if span.is_recording():
+            span.set_attribute(_TASK_TAG_KEY, _TASK_SEND)
+            span.set_attribute(SpanAttributes.MESSAGING_MESSAGE_ID, message.task_id)
+            span.set_attribute(_TASK_NAME_KEY, message.task_name)
+            set_attributes_from_context(span, message.labels)
+
+        activation = trace.use_span(span, end_on_exit=True)
```

Comment:
Here's a small quetsion about spans. It seems like you use spans as events. 

From my understanding it makes sense to only have 1 single span for task execution. And inside of that span we can use events to show what was going on during the exeuction. 

I'm not sure if it's the best way. I need to try it out.